pulsar-common/src/main/java/org/apache/pulsar/common/compression/CompressionCodecNone.java

```diff
@@ -28,8 +29,15 @@ public class CompressionCodecNone implements CompressionCodec {
 
     @Override
     public ByteBuf encode(ByteBuf raw) {
-        // Provides an encoder that simply returns the same uncompressed buffer
-        return raw.retain();
+        int readableBytes = raw.readableBytes();
+        if (readableBytes == raw.capacity()) {
+            return raw.retain();
+        }
+        raw.markReaderIndex();
+        ByteBuf target = PulsarByteBufAllocator.DEFAULT.buffer(readableBytes, readableBytes);
+        target.writeBytes(raw);
+        raw.resetReaderIndex();
+        return target;
```

Comment:
I don't think this is the right solution because it would cause a memory copy for all the batches.

Reply:
+1. 
1. We should take the whole buffer size into account in `MemoryLimitController`. Because it's allocated by our producer.
2. There is a memory waste problem in producer batch container, as the max buffer size only grows.